ibi-group/otp-middleware
# Trip monitoring: Notify at leadtime

### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR makes OTP-middleware send an initial reminder notification at the start of the trip monitoring window ("Monitor this trip... 30 mins before" under the UI advanced settings for a trip). The PR also trims messages so that they fit SMS and push notification message length limitations (otherwise sending fails).

This PR works with OTP-middleware configurations where `PUSH_API_URL` and `PUSH_API_KEY` are specified, and OTP-RR configurations where `persistence > strategy` is `otp_middleware` and `persistence > otp_middleware > supportsPushNotifications` is `true`.

To test:
1. Register your mobile device by opening the mobile app.
2. Log in from the OTP-RR UI and, under "My Account > Settings > Notifications", check "Push notifications".
3. Plan a trip on a frequent transit line departing say within the next hour (e.g. on a popular rail line)
4. Save the trip for monitoring (Check that trip availability is confirmed for the current day)
5. Wait for a push notification on your mobile device.
